File: apps/website/src/server/db/seeds/suma-de-los-dados/solution-dynamicProgramming.py
def sumaDeLosDados(cantDados, cantCaras, sumaRequerida):
    rows = cantDados + 1
    columns = sumaRequerida + 1
    matrix = [[None for _ in range(columns)] for _ in range(rows)]
    return sumaParcial(matrix, cantCaras, cantDados, sumaRequerida)


def sumaParcial(matrix, cantCaras, cantDados, sumaRequerida):
    if cantDados == 1:
        if sumaRequerida >= 1 and sumaRequerida <= cantCaras:
            return 1
        return 0
    if matrix[cantDados][sumaRequerida] is None:
        suma = 0
        minimo = min(cantCaras, sumaRequerida)
        for i in range(1, minimo + 1):
            suma += sumaParcial(matrix, cantCaras, cantDados - 1, sumaRequerida - i)
        matrix[cantDados][sumaRequerida] = suma
    return matrix[cantDados][sumaRequerida]

File: apps/website/src/server/db/seeds/suma-de-los-dados/solution-dynamicProgramming.php
<?php

function sumaDeLosDados(int $cantDados, int $cantCaras, int $sumaRequerida): int {
    $rows = $cantDados + 1;
    $columns = $sumaRequerida + 1;
    $matrix = array_fill(0, $rows, array_fill(0, $columns, null));
    return sumaParcial($matrix, $cantCaras, $cantDados, $sumaRequerida);
}

function sumaParcial(array &$matrix, int $cantCaras, int $cantDados, int $sumaRequerida): int {
    if ($cantDados == 1) {
        if ($sumaRequerida >= 1 && $sumaRequerida <= $cantCaras) {
            return 1;
        }
        return 0;
    }
    if ($matrix[$cantDados][$sumaRequerida] == null) {
        $suma = 0;
        $minimo = min($cantCaras, $sumaRequerida);
        for ($i = 1; $i <= $minimo; $i++) {
            $suma += sumaParcial($matrix, $cantCaras, $cantDados - 1, $sumaRequerida - $i);
        }
        $matrix[$cantDados][$sumaRequerida] = $suma;
    }
    return $matrix[$cantDados][$sumaRequerida];
}
